src/pudl/helpers.py

```diff
@@ -332,90 +332,224 @@ def is_doi(doi):
+    right = convert_col_to_datetime(right, right_date_col)
+    left = convert_col_to_datetime(left, left_date_col)
+    date_on = [col + "_temp_for_merge" for col in date_on]
+    right = separate_date_cols(right, right_date_col, date_on)
+    left = separate_date_cols(left, left_date_col, date_on)
+    merge_cols = date_on + on
+    out = pd.merge(left, right, on=merge_cols, how=how, **kwargs)
+
+    suffixes = ["", ""]
+    if left_date_col == right_date_col:
+        if "suffixes" in kwargs:
+            suffixes = kwargs["suffixes"]
+        else:
+            suffixes = ["_x", "_y"]
+    # reconstruct the new report date column and clean up columns
+    left_right_date_col = [left_date_col + suffixes[0], right_date_col + suffixes[1]]
+    if report_at_start:
+        # keep the later of the two report dates when determining
+        # the new report date for each row
+        reconstructed_date = out[left_right_date_col].max(axis=1)
+    else:
+        # keep the earlier of the two report dates
+        reconstructed_date = out[left_right_date_col].min(axis=1)
+    out = out.drop(left_right_date_col + date_on, axis=1)
+    out.insert(loc=0, column=new_date_col, value=reconstructed_date)
+    return out
+
+
+def expand_timeseries(
+    df: pd.DataFrame,
+    date_col: str = "report_date",
+    freq: str = "MS",
+    key_cols: List[str] = [],
+) -> pd.DataFrame:
+    """Expand a dataframe to a include a full time series at a given frequency.
+
+    This function adds a full timeseries to the given dataframe for each group
+    of columns specified by ``key_cols``. The data in the timeseries will be filled
+    with the next previous chronological observation for a group of primary key columns
+    specified by ``key_cols``.
+
+    Arguments:
+        df: The dataframe to expand. Must have ``date_col`` in columns.
+        date_col: Name of the datetime column being expanded into a full timeseries.
+        freq: The frequency of the time series to expand the data to.
+            See :ref:`here <timeseries.offset_aliases>` for a list of
+            frequency aliases.
+        key_cols: Column names of the non-date primary key columns in the dataframe.
+            The resulting dataframe will have a full timeseries expanded for each
+            unique group of these ID columns that are present in the dataframe.
+    """
+    try:
+        pd.tseries.frequencies.to_offset(freq)
+    except ValueError:
+        logger.exception(
+            f"Frequency string {freq} is not valid. \
+            See Pandas Timeseries Offset Aliases docs for valid strings."
+        )
+    # for each group of ID columns add a dummy record with the date column
+    # equal to Jan 1 of the year after the last real record in the group
+    # this allows records to be filled through the end of the last reported year
+    end_dates = df.groupby(key_cols).agg({date_col: "max"})
+    end_dates.loc[:, date_col] = end_dates[date_col].dt.year + 1
```

Comment:
Hmm.  There should be some test cases for this probably. We only want to fill in records to the end of the last reported year if we're filling in the time series at a higher frequency. Filling in values for additional years that weren't part of the dataset at all seems like it would cause problems downstream.

Reply:
I feel a little confused about desired behavior here. Currently this fills in through the end of the last reported year (see the unit test `test_timeseries_fillin` where monthly data is filled in through December of the last reported year). The only way I could figure out how to do this is to add a dummy row for Jan 1 of the following year. For example:
- last reported value is 2/1/2019
- add dummy row for 1/1/2020
- function fills in monthly values from 2/1/2019 to 12/1/2019
- the dummy row for 1/1/2020 is dropped

It definitely seems like a weird way to do this, but I also couldn't find a nice way to do it so maybe it really is the best way.

 I guess in theory data could be filled in to the end of a specified frequency. So for example if data was daily there could be an argument passed in that specifies that data should be filled in to the end of the month instead of the end of the year.